From 857e26a395a446da7b6bb3a1056ef3fb8d85cf65 Mon Sep 17 00:00:00 2001
From: majik-sheff <tim@tfthorpe.net>
Date: Mon, 25 Sep 2017 04:38:49 -0500
Subject: [PATCH] ar71xx: add support for TL-WR1043N v5.0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The hardware of the TP-Link WR-1043N v5 appears
to be to almost identical to the WR-1043NDv4, except
that the USB port has been removed and there is no
longer a removeable antenna option.

The software is more in line with the Archer series in
that is uses a nested bootloader scheme. The factory
image should install from the stock interface.

Set your Ethernet adapter to 192.168.0.10. Enable your TFTP
server with the firmware image in the root named fw.bin
(can be any name, just change the tftp command that comes later)
When the router starts its boot process,
press “enter” to interrupt auto boot.

Enter the following commands:
tftp 0x80060000 fw.bin
erase 0x9f040000 +${filesize}
cp.b 0x80060000 0x9f040000 ${filesize}
reset

Signed-off-by: Tim Thorpe <tim@tfthorpe.net>
[Rebased on 8ad1b09, Fixed TPLINK_BOARD_NAME]
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
---
 target/linux/ar71xx/base-files/etc/board.d/01_leds |  10 +-
 .../linux/ar71xx/base-files/etc/board.d/02_network |   4 +
 target/linux/ar71xx/base-files/etc/diag.sh         |   1 +
 .../etc/hotplug.d/ieee80211/10_fix_wifi_mac        |  24 ++
 target/linux/ar71xx/base-files/lib/ar71xx.sh       |   3 +
 .../ar71xx/base-files/lib/upgrade/platform.sh      |   2 +
 target/linux/ar71xx/config-4.4                     |   1 +
 .../ar71xx/files/arch/mips/ath79/Kconfig.openwrt   |  10 +
 .../files/arch/mips/ath79/mach-tl-wr1043nd-v4.c    | 316 +++++++++++++--------
 .../linux/ar71xx/files/arch/mips/ath79/machtypes.h |   1 +
 target/linux/ar71xx/image/tp-link.mk               |  17 +-
 target/linux/ar71xx/mikrotik/config-default        |   1 +
 target/linux/ar71xx/nand/config-default            |   1 +
 tools/firmware-utils/src/tplink-safeloader.c       |  39 ++-
 14 files changed, 310 insertions(+), 120 deletions(-)
 create mode 100644 target/linux/ar71xx/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac

diff --git a/target/linux/ar71xx/base-files/etc/board.d/01_leds b/target/linux/ar71xx/base-files/etc/board.d/01_leds
index e1efb56..5f58701 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/01_leds
+++ b/target/linux/ar71xx/base-files/etc/board.d/01_leds
@@ -666,14 +666,20 @@ tl-wr1043nd-v2)
 	ucidef_set_led_usbdev "usb" "USB" "tp-link:green:usb" "1-1"
 	ucidef_set_led_wlan "wlan" "WLAN" "tp-link:green:wlan" "phy0tpt"
 	;;
-tl-wr1043nd-v4)
-	ucidef_set_led_usbdev "usb" "USB" "tp-link:green:usb" "1-1"
+tl-wr1043nd-v4|\
+tl-wr1043n-v5)
 	ucidef_set_led_wlan "wlan" "WLAN" "tp-link:green:wlan" "phy0tpt"
 	ucidef_set_led_switch "wan" "WAN" "tp-link:green:wan" "switch0" "0x20"
 	ucidef_set_led_switch "lan1" "LAN1" "tp-link:green:lan1" "switch0" "0x10"
 	ucidef_set_led_switch "lan2" "LAN2" "tp-link:green:lan2" "switch0" "0x08"
 	ucidef_set_led_switch "lan3" "LAN3" "tp-link:green:lan3" "switch0" "0x04"
 	ucidef_set_led_switch "lan4" "LAN4" "tp-link:green:lan4" "switch0" "0x02"
+
+	case "$board" in
+		tl-wr1043nd-v4)
+		ucidef_set_led_usbdev "usb" "USB" "tp-link:green:usb" "1-1"
+		;;
+	esac
 	;;
 tl-wr2543n)
 	ucidef_set_led_usbdev "usb" "USB" "tp-link:green:usb" "1-1"
diff --git a/target/linux/ar71xx/base-files/etc/board.d/02_network b/target/linux/ar71xx/base-files/etc/board.d/02_network
index 24ead87..e7a1702 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/02_network
+++ b/target/linux/ar71xx/base-files/etc/board.d/02_network
@@ -235,6 +235,7 @@ ar71xx_setup_interfaces()
 	dir-835-a1|\
 	esr900|\
 	mynet-n750|\
+	tl-wr1043n-v5|\
 	sr3200|\
 	wndr3700v4|\
 	wndr4300)
@@ -485,6 +486,9 @@ ar71xx_setup_macs()
 		lan_mac=$(mtd_get_mac_binary product-info 8)
 		wan_mac=$(macaddr_add "$lan_mac" 1)
 		;;
+	tl-wr1043n-v5)
+		wan_mac=$(macaddr_add $(mtd_get_mac_binary product-info 8) 1)
+		;;	
 	esr900)
 		wan_mac=$(mtd_get_mac_ascii u-boot-env "wanaddr")
 		;;
diff --git a/target/linux/ar71xx/base-files/etc/diag.sh b/target/linux/ar71xx/base-files/etc/diag.sh
index 38cc5d7..ec0c720 100644
--- a/target/linux/ar71xx/base-files/etc/diag.sh
+++ b/target/linux/ar71xx/base-files/etc/diag.sh
@@ -362,6 +362,7 @@ get_status_led() {
 	tl-wr1043nd|\
 	tl-wr1043nd-v2|\
 	tl-wr1043nd-v4|\
+	tl-wr1043n-v5|\
 	tl-wr741nd|\
 	tl-wr741nd-v4|\
 	tl-wa801nd-v3|\
diff --git a/target/linux/ar71xx/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac b/target/linux/ar71xx/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
new file mode 100644
index 0000000..4b0922f
--- /dev/null
+++ b/target/linux/ar71xx/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
@@ -0,0 +1,24 @@
+#!/bin/ash
+
+[ "$ACTION" == "add" ] || exit 0
+
+PHYNBR=${DEVPATH##*/phy}
+
+[ -n $PHYNBR ] || exit 0
+
+. /lib/functions.sh
+. /lib/functions/system.sh
+
+board=$(board_name)
+
+case "$board" in
+	archer-c58-v1)
+		echo $(macaddr_add $(mtd_get_mac_binary mac 8)  $(($PHYNBR - 1)) ) > /sys${DEVPATH}/macaddress
+		;;
+	tl-wr1043n-v5)
+		echo $(macaddr_add $(mtd_get_mac_binary product-info 8)  $(($PHYNBR - 1)) ) > /sys${DEVPATH}/macaddress
+		;;
+	*)
+		;;
+esac
+
diff --git a/target/linux/ar71xx/base-files/lib/ar71xx.sh b/target/linux/ar71xx/base-files/lib/ar71xx.sh
index 46711af..de6042b 100755
--- a/target/linux/ar71xx/base-files/lib/ar71xx.sh
+++ b/target/linux/ar71xx/base-files/lib/ar71xx.sh
@@ -944,6 +944,9 @@ ar71xx_board_detect() {
 	*"TL-WR1043ND v4")
 		name="tl-wr1043nd-v4"
 		;;
+	*"TL-WR1043N v5")
+		name="tl-wr1043n-v5"
+		;;
 	*TL-WR2543N*)
 		name="tl-wr2543n"
 		;;
diff --git a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
index e65f6e2..ec8e989 100755
--- a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
@@ -261,6 +261,7 @@ platform_check_image() {
 	tew-712br|\
 	tew-732br|\
 	tew-823dru|\
+	tl-wr1043n-v5|\
 	unifi-outdoor|\
 	unifiac-lite|\
 	unifiac-pro|\
@@ -291,6 +292,7 @@ platform_check_image() {
 		;;
 	alfa-ap96|\
 	alfa-nx|\
+	tl-wr1043n-v5|\
 	ap121-mini|\
 	ap121|\
 	ap135-020|\
diff --git a/target/linux/ar71xx/config-4.4 b/target/linux/ar71xx/config-4.4
index 0187775..4cf0c6f 100644
--- a/target/linux/ar71xx/config-4.4
+++ b/target/linux/ar71xx/config-4.4
@@ -181,6 +181,7 @@ CONFIG_ATH79_MACH_TL_WR1041N_V2=y
 CONFIG_ATH79_MACH_TL_WR1043ND=y
 CONFIG_ATH79_MACH_TL_WR1043ND_V2=y
 CONFIG_ATH79_MACH_TL_WR1043ND_V4=y
+CONFIG_ATH79_MACH_TL_WR1043N_V5=y
 CONFIG_ATH79_MACH_TL_WR2543N=y
 CONFIG_ATH79_MACH_TL_WR703N=y
 CONFIG_ATH79_MACH_TL_WR720N_V3=y
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt b/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
index 4112fd3..94f25ee 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
+++ b/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
@@ -1596,6 +1596,16 @@ config ATH79_MACH_TL_WR1043ND_V4
 	select ATH79_DEV_USB
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_TL_WR1043N_V5
+	bool "TP-LINK TL-WR1043N v5 support"
+	select SOC_QCA956X
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_SPI
+	select ATH79_DEV_WMAC
+	select ATH79_DEV_ETH
+	select ATH79_DEV_M25P80
+
 config ATH79_MACH_TL_WR2543N
 	bool "TP-LINK TL-WR2543N/ND support"
 	select SOC_AR724X
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043nd-v4.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043nd-v4.c
index b1539c5..53a82d2 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043nd-v4.c
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043nd-v4.c
@@ -5,6 +5,7 @@
  *  Copyright (C) 2016 Matthias Schiffer <mschiffer@universe-factory.net>
  *  Copyright (C) 2016 Andreas Ziegler <github@andreas-ziegler.de>
  *  Copyright (C) 2016 Ludwig Thomeczek <ledesrc@wxorx.net>
+ *  Copyright (c) 2017 Tim Thorpe <tim@tfthorpe.net>
  *
  *  Derived from: mach-dir-869-a1.c
  *
@@ -38,153 +39,236 @@
 #include "machtypes.h"
 #include "nvram.h"
 
-#define TL_WR1043_V4_GPIO_BTN_RESET		2
-#define TL_WR1043_V4_GPIO_BTN_RFKILL		5
+#define TL_WR1043_V4_GPIO_BTN_RESET       2
+#define TL_WR1043_V4_GPIO_BTN_RFKILL      5
 
-#define TL_WR1043_V4_GPIO_LED_WLAN		19
-#define TL_WR1043_V4_GPIO_LED_USB		7
-#define TL_WR1043_V4_GPIO_LED_WPS		1
-#define TL_WR1043_V4_GPIO_LED_SYSTEM		6
+#define TL_WR1043_V4_GPIO_LED_WLAN       19
+#define TL_WR1043_V4_GPIO_LED_USB         7
+#define TL_WR1043_V4_GPIO_LED_WPS         1
+#define TL_WR1043_V4_GPIO_LED_SYSTEM      6
 
-#define TL_WR1043_V4_GPIO_USB_POWER		8
+#define TL_WR1043_V4_GPIO_USB_POWER       8
 
-#define TL_WR1043_V4_GPIO_LED_WAN		15
-#define TL_WR1043_V4_GPIO_LED_LAN1		9
-#define TL_WR1043_V4_GPIO_LED_LAN2		14
-#define TL_WR1043_V4_GPIO_LED_LAN3		21
-#define TL_WR1043_V4_GPIO_LED_LAN4		20
+#define TL_WR1043_V4_GPIO_LED_WAN        15
+#define TL_WR1043_V5_GPIO_LED_WAN_ORANGE 16
+#define TL_WR1043_V4_GPIO_LED_LAN1        9
+#define TL_WR1043_V4_GPIO_LED_LAN2       14
+#define TL_WR1043_V4_GPIO_LED_LAN3       21
+#define TL_WR1043_V4_GPIO_LED_LAN4       20
 
-#define TL_WR1043_V4_KEYS_POLL_INTERVAL		20 /* msecs */
-#define TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL	(3 * TL_WR1043_V4_KEYS_POLL_INTERVAL)
+#define TL_WR1043_V4_KEYS_POLL_INTERVAL      20 /* msecs */
+#define TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL (3 * TL_WR1043_V4_KEYS_POLL_INTERVAL)
 
-#define TL_WR1043_V4_MAC_LOCATION		0x1ff50008
-
-#define TL_WR1043_V4_EEPROM_ADDR		0x1fff0000
-#define TL_WR1043_V4_WMAC_CALDATA_OFFSET	0x1000
+#define TL_WR1043_V4_MAC_LOCATION        0x1ff50008
+#define TL_WR1043_V4_EEPROM_ADDR         0x1fff0000
+#define TL_WR1043_V4_WMAC_CALDATA_OFFSET     0x1000
+#define TL_WR1043_V5_MAC_LOCATION        0x1ff00008
+#define TL_WR1043_V5_DEF_PIN_LOCATION    0x1ff00208
+#define TL_WR1043_V5_TPLINK_XML_LOCATION 0x1ff1c008
 
 static struct gpio_led tl_wr1043nd_v4_leds_gpio[] __initdata = {
-	{
-		.name		= "tp-link:green:wps",
-		.gpio		= TL_WR1043_V4_GPIO_LED_WPS,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:system",
-		.gpio		= TL_WR1043_V4_GPIO_LED_SYSTEM,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:wlan",
-		.gpio		= TL_WR1043_V4_GPIO_LED_WLAN,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:usb",
-		.gpio		= TL_WR1043_V4_GPIO_LED_USB,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:wan",
-		.gpio		= TL_WR1043_V4_GPIO_LED_WAN,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:lan1",
-		.gpio		= TL_WR1043_V4_GPIO_LED_LAN1,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:lan2",
-		.gpio		= TL_WR1043_V4_GPIO_LED_LAN2,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:lan3",
-		.gpio		= TL_WR1043_V4_GPIO_LED_LAN3,
-		.active_low	= 1,
-	},
-	{
-		.name		= "tp-link:green:lan4",
-		.gpio		= TL_WR1043_V4_GPIO_LED_LAN4,
-		.active_low	= 1,
-	},
+  {
+    .name        = "tp-link:green:wps",
+    .gpio        = TL_WR1043_V4_GPIO_LED_WPS,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:system",
+    .gpio        = TL_WR1043_V4_GPIO_LED_SYSTEM,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:wlan",
+    .gpio        = TL_WR1043_V4_GPIO_LED_WLAN,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:usb",
+    .gpio        = TL_WR1043_V4_GPIO_LED_USB,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:wan",
+    .gpio        = TL_WR1043_V4_GPIO_LED_WAN,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan1",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN1,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan2",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN2,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan3",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN3,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan4",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN4,
+    .active_low  = 1,
+  },
 };
 
 static struct gpio_keys_button tl_wr1043nd_v4_gpio_keys[] __initdata = {
-	{
-		.desc		= "Reset button",
-		.type		= EV_KEY,
-		.code		= KEY_RESTART,
-		.debounce_interval = TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL,
-		.gpio		= TL_WR1043_V4_GPIO_BTN_RESET,
-		.active_low	= 1,
-	},
-	{
-		.desc		= "RFKILL button",
-		.type		= EV_KEY,
-		.code		= KEY_RFKILL,
-		.debounce_interval = TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL,
-		.gpio		= TL_WR1043_V4_GPIO_BTN_RFKILL,
-		.active_low	= 1,
-	},
+  {
+    .desc              = "Reset button",
+    .type              = EV_KEY,
+    .code              = KEY_RESTART,
+    .debounce_interval = TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL,
+    .gpio              = TL_WR1043_V4_GPIO_BTN_RESET,
+    .active_low        = 1,
+  },
+  {
+    .desc              = "RFKILL button",
+    .type              = EV_KEY,
+    .code              = KEY_RFKILL,
+    .debounce_interval = TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL,
+    .gpio              = TL_WR1043_V4_GPIO_BTN_RFKILL,
+    .active_low        = 1,
+  },
 };
 
 static struct ar8327_pad_cfg tl_wr1043nd_v4_ar8327_pad0_cfg = {
-	.mode = AR8327_PAD_MAC_SGMII,
-	.sgmii_delay_en = true,
+  .mode           = AR8327_PAD_MAC_SGMII,
+  .sgmii_delay_en = true,
 };
 
 static struct ar8327_platform_data tl_wr1043nd_v4_ar8327_data = {
-	.pad0_cfg = &tl_wr1043nd_v4_ar8327_pad0_cfg,
-	.port0_cfg = {
-		.force_link = 1,
-		.speed = AR8327_PORT_SPEED_1000,
-		.duplex = 1,
-		.txpause = 1,
-		.rxpause = 1,
-	},
+  .pad0_cfg = &tl_wr1043nd_v4_ar8327_pad0_cfg,
+  .port0_cfg = {
+    .force_link = 1,
+    .speed      = AR8327_PORT_SPEED_1000,
+    .duplex     = 1,
+    .txpause    = 1,
+    .rxpause    = 1,
+  },
 };
 
 static struct mdio_board_info tl_wr1043nd_v4_mdio0_info[] = {
-	{
-		.bus_id = "ag71xx-mdio.0",
-		.phy_addr = 0,
-		.platform_data = &tl_wr1043nd_v4_ar8327_data,
-	},
+  {
+    .bus_id        = "ag71xx-mdio.0",
+    .phy_addr      = 0,
+    .platform_data = &tl_wr1043nd_v4_ar8327_data,
+  },
 };
 
 static void __init tl_wr1043nd_v4_setup(void)
 {
-	u8 *mac = (u8 *) KSEG1ADDR(TL_WR1043_V4_MAC_LOCATION);
-	u8 *eeprom = (u8 *) KSEG1ADDR(TL_WR1043_V4_EEPROM_ADDR);
+  u8 *mac = (u8 *) KSEG1ADDR(TL_WR1043_V4_MAC_LOCATION);
+  u8 *eeprom = (u8 *) KSEG1ADDR(TL_WR1043_V4_EEPROM_ADDR);
 
-	ath79_register_m25p80(NULL);
+  ath79_register_m25p80(NULL);
 
-	ath79_init_mac(ath79_eth0_data.mac_addr, mac, 0);
-	ath79_eth0_data.phy_if_mode = PHY_INTERFACE_MODE_SGMII;
-	ath79_eth0_data.mii_bus_dev = &ath79_mdio0_device.dev;
-	ath79_eth0_data.phy_mask = BIT(0);
+  ath79_init_mac(ath79_eth0_data.mac_addr, mac, 0);
+  ath79_eth0_data.phy_if_mode = PHY_INTERFACE_MODE_SGMII;
+  ath79_eth0_data.mii_bus_dev = &ath79_mdio0_device.dev;
+  ath79_eth0_data.phy_mask = BIT(0);
 
-	mdiobus_register_board_info(tl_wr1043nd_v4_mdio0_info,
-	                            ARRAY_SIZE(tl_wr1043nd_v4_mdio0_info));
+  mdiobus_register_board_info(tl_wr1043nd_v4_mdio0_info,
+                             ARRAY_SIZE(tl_wr1043nd_v4_mdio0_info));
 
-	ath79_register_usb();
-	ath79_register_mdio(0, 0);
-	ath79_register_eth(0);
+  ath79_register_usb();
+  ath79_register_mdio(0, 0);
+  ath79_register_eth(0);
 
-	ath79_register_wmac(eeprom + TL_WR1043_V4_WMAC_CALDATA_OFFSET, mac);
+  ath79_register_wmac(eeprom + TL_WR1043_V4_WMAC_CALDATA_OFFSET, mac);
 
-	ath79_register_leds_gpio(-1, ARRAY_SIZE(tl_wr1043nd_v4_leds_gpio),
-	                         tl_wr1043nd_v4_leds_gpio);
+  ath79_register_leds_gpio(-1, ARRAY_SIZE(tl_wr1043nd_v4_leds_gpio),
+                          tl_wr1043nd_v4_leds_gpio);
 
-	ath79_register_gpio_keys_polled(-1, TL_WR1043_V4_KEYS_POLL_INTERVAL,
-	                                ARRAY_SIZE(tl_wr1043nd_v4_gpio_keys),
-	                                tl_wr1043nd_v4_gpio_keys);
+  ath79_register_gpio_keys_polled(-1, TL_WR1043_V4_KEYS_POLL_INTERVAL,
+                                 ARRAY_SIZE(tl_wr1043nd_v4_gpio_keys),
+                                 tl_wr1043nd_v4_gpio_keys);
 
-	gpio_request_one(TL_WR1043_V4_GPIO_USB_POWER,
-			 GPIOF_OUT_INIT_HIGH | GPIOF_EXPORT_DIR_FIXED,
-			 "USB power");
+  gpio_request_one(TL_WR1043_V4_GPIO_USB_POWER,
+                  GPIOF_OUT_INIT_HIGH | GPIOF_EXPORT_DIR_FIXED,
+                  "USB power");
 }
 
 MIPS_MACHINE(ATH79_MACH_TL_WR1043ND_V4, "TL-WR1043ND-v4",
-	     "TP-LINK TL-WR1043ND v4", tl_wr1043nd_v4_setup);
+            "TP-LINK TL-WR1043ND v4", tl_wr1043nd_v4_setup);
+
+static struct gpio_led tl_wr1043n_v5_leds_gpio[] __initdata = {
+  {
+    .name        = "tp-link:green:system",
+    .gpio        = TL_WR1043_V4_GPIO_LED_SYSTEM,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:wlan",
+    .gpio        = TL_WR1043_V4_GPIO_LED_WLAN,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:wan",
+    .gpio        = TL_WR1043_V4_GPIO_LED_WAN,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:orange:wan",
+    .gpio        = TL_WR1043_V5_GPIO_LED_WAN_ORANGE,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan1",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN1,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan2",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN2,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan3",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN3,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:lan4",
+    .gpio        = TL_WR1043_V4_GPIO_LED_LAN4,
+    .active_low  = 1,
+  },
+  {
+    .name        = "tp-link:green:wps",
+    .gpio        = TL_WR1043_V4_GPIO_LED_WPS,
+    .active_low  = 1,
+  },
+};
+
+/* The 1043Nv5 is identical to the 1043NDv4, only missing the usb and small firmware layout changes  */
+static void __init tl_wr1043nv5_setup(void)
+{
+  u8 *mac = (u8 *) KSEG1ADDR(TL_WR1043_V5_MAC_LOCATION);
+  u8 *eeprom = (u8 *) KSEG1ADDR(TL_WR1043_V4_EEPROM_ADDR);
+
+  ath79_register_m25p80(NULL);
+
+  ath79_init_mac(ath79_eth0_data.mac_addr, mac, 0);
+  ath79_eth0_data.phy_if_mode = PHY_INTERFACE_MODE_SGMII;
+  ath79_eth0_data.mii_bus_dev = &ath79_mdio0_device.dev;
+  ath79_eth0_data.phy_mask = BIT(0);
+
+  mdiobus_register_board_info(tl_wr1043nd_v4_mdio0_info,
+                             ARRAY_SIZE(tl_wr1043nd_v4_mdio0_info));
+
+  ath79_register_mdio(0, 0);
+  ath79_register_eth(0);
+
+  ath79_register_wmac(eeprom + TL_WR1043_V4_WMAC_CALDATA_OFFSET, mac);
+
+  ath79_register_leds_gpio(-1, ARRAY_SIZE(tl_wr1043n_v5_leds_gpio),
+                          tl_wr1043n_v5_leds_gpio);
+
+  ath79_register_gpio_keys_polled(-1, TL_WR1043_V4_KEYS_POLL_INTERVAL,
+                                 ARRAY_SIZE(tl_wr1043nd_v4_gpio_keys),
+                                 tl_wr1043nd_v4_gpio_keys);
+}
+
+MIPS_MACHINE(ATH79_MACH_TL_WR1043N_V5, "TL-WR1043N-v5", "TP-LINK TL-WR1043N v5",
+            tl_wr1043nv5_setup);
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h b/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
index 8864e0d..11fc236 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
+++ b/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
@@ -217,6 +217,7 @@ enum ath79_mach_type {
 	ATH79_MACH_TL_WR1043ND,			/* TP-LINK TL-WR1043ND */
 	ATH79_MACH_TL_WR1043ND_V2,		/* TP-LINK TL-WR1043ND v2 */
 	ATH79_MACH_TL_WR1043ND_V4,		/* TP-LINK TL-WR1043ND v4 */
+	ATH79_MACH_TL_WR1043N_V5,		/* TP-LINK TL-WR1043N v5 */
 	ATH79_MACH_TL_WR2543N,			/* TP-LINK TL-WR2543N/ND */
 	ATH79_MACH_TL_WR703N,			/* TP-LINK TL-WR703N */
 	ATH79_MACH_TL_WR710N,			/* TP-LINK TL-WR710N */
diff --git a/target/linux/ar71xx/image/tp-link.mk b/target/linux/ar71xx/image/tp-link.mk
index 685b09e..61ebb09 100644
--- a/target/linux/ar71xx/image/tp-link.mk
+++ b/target/linux/ar71xx/image/tp-link.mk
@@ -843,7 +843,22 @@ define Device/tl-wr1043nd-v4
     IMAGE/sysupgrade.bin := append-rootfs | tplink-safeloader sysupgrade
     IMAGE/factory.bin := append-rootfs | tplink-safeloader factory
 endef
-TARGET_DEVICES += tl-wr1043nd-v1 tl-wr1043nd-v2 tl-wr1043nd-v3 tl-wr1043nd-v4
+define Device/tl-wr1043n-v5
+  $(Device/tplink)
+  DEVICE_TITLE := TP-LINK TL-WR1043N v5
+  BOARDNAME := TL-WR1043N-v5
+  DEVICE_PROFILE := TLWR1043
+  TPLINK_HWID :=  0x10430005
+  MTDPARTS := spi0.0:128k(factory-uboot)ro,128k(u-boot)ro,1536k(kernel),13568k(rootfs),128k(product-info)ro,640k(config)ro,64k(partition-table)ro,128k(logs)ro,64k(ART)ro,15104k@0x40000(firmware)
+  IMAGE_SIZE := 15104k
+  TPLINK_BOARD_NAME := TLWR1043NV5
+  KERNEL := kernel-bin | patch-cmdline | lzma | uImageArcher lzma
+  IMAGES := sysupgrade.bin factory.bin
+  IMAGE/sysupgrade.bin := append-rootfs | tplink-safeloader sysupgrade | \
+	append-metadata | check-size $$$$(IMAGE_SIZE)
+  IMAGE/factory.bin := append-rootfs | tplink-safeloader factory
+endef
+TARGET_DEVICES += tl-wr1043nd-v1 tl-wr1043nd-v2 tl-wr1043nd-v3 tl-wr1043nd-v4 tl-wr1043n-v5
 
 define Device/tl-wr2543-v1
     $(Device/tplink-8mlzma)
diff --git a/target/linux/ar71xx/mikrotik/config-default b/target/linux/ar71xx/mikrotik/config-default
index 376835a..878f7e8 100644
--- a/target/linux/ar71xx/mikrotik/config-default
+++ b/target/linux/ar71xx/mikrotik/config-default
@@ -146,6 +146,7 @@ CONFIG_ATH79_MACH_RBSXTLITE=y
 # CONFIG_ATH79_MACH_TL_WR1043ND is not set
 # CONFIG_ATH79_MACH_TL_WR1043ND_V2 is not set
 # CONFIG_ATH79_MACH_TL_WR1043ND_V4 is not set
+# CONFIG_ATH79_MACH_TL_WR1043N_V5 is not set
 # CONFIG_ATH79_MACH_TL_WR2543N is not set
 # CONFIG_ATH79_MACH_TL_WR703N is not set
 # CONFIG_ATH79_MACH_TL_WR720N_V3 is not set
diff --git a/target/linux/ar71xx/nand/config-default b/target/linux/ar71xx/nand/config-default
index 62be218..0ccbb4c 100644
--- a/target/linux/ar71xx/nand/config-default
+++ b/target/linux/ar71xx/nand/config-default
@@ -54,6 +54,7 @@
 # CONFIG_ATH79_MACH_TL_WR1041N_V2 is not set
 # CONFIG_ATH79_MACH_TL_WR1043ND is not set
 # CONFIG_ATH79_MACH_TL_WR1043ND_V4 is not set
+# CONFIG_ATH79_MACH_TL_WR1043N_V5 is not set
 # CONFIG_ATH79_MACH_TL_WR2543N is not set
 # CONFIG_ATH79_MACH_TL_WR703N is not set
 # CONFIG_ATH79_MACH_TL_WR720N_V3 is not set
diff --git a/tools/firmware-utils/src/tplink-safeloader.c b/tools/firmware-utils/src/tplink-safeloader.c
index 2468426..72263a4 100644
--- a/tools/firmware-utils/src/tplink-safeloader.c
+++ b/tools/firmware-utils/src/tplink-safeloader.c
@@ -478,6 +478,42 @@ static struct device_info boards[] = {
 		.last_sysupgrade_partition = "file-system"
 	},
 
+	/** Firmware layout for the TL-WR1043 v5 */
+	{
+		.id     = "TLWR1043NV5",
+		.vendor = "",
+		.support_list =
+			"SupportList:\n"
+			"{product_name:TL-WR1043N,product_ver:5.0.0,special_id:45550000}\n"
+			"{product_name:TL-WR1043N,product_ver:5.0.0,special_id:55530000}\n",
+		.support_trail = '\x00',
+		.soft_ver = "soft_ver:1.0.0\n",
+		.partitions = {
+			{"factory-boot", 0x00000, 0x20000},
+			{"fs-uboot", 0x20000, 0x20000},
+			{"os-image", 0x40000, 0x180000},
+			{"file-system", 0x1c0000, 0xd40000},
+			{"default-mac", 0xf00000, 0x00200},
+			{"pin", 0xf00200, 0x00200},
+			{"device-id", 0xf00400, 0x00100},
+			{"product-info", 0xf00500, 0x0fb00},
+			{"soft-version", 0xf10000, 0x01000},
+			{"extra-para", 0xf11000, 0x01000},
+			{"support-list", 0xf12000, 0x0a000},
+			{"profile", 0xf1c000, 0x04000},
+			{"default-config", 0xf20000, 0x10000},
+			{"user-config", 0xf30000, 0x40000},
+			{"qos-db", 0xf70000, 0x40000},
+			{"certificate", 0xfb0000, 0x10000},
+			{"partition-table", 0xfc0000, 0x10000},
+			{"log", 0xfd0000, 0x20000},
+			{"radio", 0xff0000, 0x10000},
+			{NULL, 0, 0}
+		},
+		.first_sysupgrade_partition = "os-image",
+		.last_sysupgrade_partition = "file-system"
+	},
+
 	/** Firmware layout for the RE450 */
 	{
 		.id = "RE450",
@@ -888,7 +924,8 @@ static void build_image(const char *output,
 	parts[3] = read_file("os-image", kernel_image, false);
 	parts[4] = read_file("file-system", rootfs_image, add_jffs2_eof);
 
-	if (strcasecmp(info->id, "ARCHER-C25-V1") == 0) {
+	/* Some devices need the extra-para partition to accept the firmware */
+	if (strcasecmp(info->id, "ARCHER-C25-V1") == 0 || strcasecmp(info->id, "TLWR1043NV5") == 0) {
 		const char mdat[11] = {0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00};
 		parts[5] = put_data("extra-para", mdat, 11);
 	}
-- 
2.7.4

